pooya changes (#246)
QuantEcon/lecture-source-jl: rst_files/julia_by_example.rst

Overview
==================

We're now ready to start learning the Julia language itself 

Level
-------

Our approach is aimed at those who already have at least some knowledge of programming --- perhaps experience with Python, MATLAB, Fortran, C or similar

> In particular, we assume you have some familiarity with fundamental programming concepts such as:

Reviewer: We can't add on colons, it is not the style of the lectures.

Reviewer 2: ```suggestion
In particular, we assume you have some familiarity with fundamental programming concepts such as
```

Set Up
--------

We assume that you've worked your way through [our getting started lecture](getting_started.ipynb) already

> In particular, the easiest way to install and precompile all of the Julia packages used in the QuantEcon notes is to go ``] add InstantiateFromURL`` and then work in a Jupyter notebook, as described [here](getting_started.ipynb)

Reviewer: I think this is markdown rather than RST links

Reviewer 2: ```suggestion
In particular, the easiest way to install and precompile all of the Julia packages used in the QuantEcon notes is to go ``] add InstantiateFromURL`` and then work in a Jupyter notebook, as described ` here <getting_started>`_
```

#. directly ``add`` the packages directly into your global installation (e.g. ``Pkg.add("MyPackage")`` or ``] add MyPackage``)  
#. download an ``Project.toml`` and ``Manifest.toml`` file in the same directory as the notebook (i.e. from the ``@__DIR__`` argument), and then call ``using Pkg; Pkg.activate(@__DIR__);``

> #. use the ``InstantiateFromURL`` package (like below):

Reviewer: I am not sure we need this comment.

#. directly ``add`` the packages directly into your global installation (e.g. ``Pkg.add("MyPackage")`` or ``] add MyPackage``)  
#. download an ``Project.toml`` and ``Manifest.toml`` file in the same directory as the notebook (i.e. from the ``@__DIR__`` argument), and then call ``using Pkg; Pkg.activate(@__DIR__);``

> #. use the ``InstantiateFromURL`` package (like below):

Reviewer 2: ```suggestion
#. use the ``InstantiateFromURL`` package
```

The arguments to ``plot`` are the numbers ``1,2, ..., n`` for the x-axis, a vector ``ϵ`` for the y-axis, and (optional) settings

> The function ``randn(n)`` returns a column vector with *n* random draws from a normal distribution mean 0 and variance 1

Reviewer: That is markdown or something similar.

Reviewer 2: The Markdown and rST syntax actually agree on this one (italics)

The arguments to ``plot`` are the numbers ``1,2, ..., n`` for the x-axis, a vector ``ϵ`` for the y-axis, and (optional) settings

The function ``randn(n)`` returns a column vector with *n* random draws from a normal distribution mean 0 and variance 1

Arrays
--------

As a language intended for mathematical and scientific computing, Julia has strong support for using symbols in the source code

In the above case, the *ϵ* and many other symbols can be typed in most Julia editor by providing the LaTeX + ``<TAB>``, i.e. ``\epsilon<TAB>`` 

> The type of a symbole is one of the most fundamental Julia data types: an array

Reviewer 2: ```suggestion
The type of a symbol is one of the most fundamental Julia data types: an array
```

The ``ϵ[1:5]`` returns an array of the first 5 elements of ``ϵ`` 

> Notice from the above that:

Reviewer 2: ```suggestion
Notice from the above that
```

---------------

Although there's no need in terms of what we wanted to achieve with our
previous code, for the sake of learning syntax let's rewrite our program to use a

> ``for`` loop for generating the data. But before that:

Reviewer 2: ```suggestion
``for`` loop for generating the data. But before that
```

.. note::
    
    In the current version of Julia v1.0, the rules for variables accessed in ``for`` and ``while`` loops can be sensitive to how they are used (and variables can sometimes require a ``global`` as part of the declaration).  We strongly advise you to avoid top level (i.e. in the `REPL <https://docs.julialang.org/en/v1/stdlib/REPL/index.html>`_ or outside of functions) ``for`` and ``while`` loops outside of Jupyter notebooks.  This issue does not apply when used within functions. For More Information go `here <https://discourse.julialang.org/t/repl-and-for-loops-scope-behavior-change/13514>`_

> Starting with the most direct version, and pretending we are in a world where `randn` can only return a single value:

Reviewer 2: ```suggestion
Starting with the most direct version, and pretending we are in a world where `randn` can only return a single value
```

The word ``in`` from the ``for`` loop can be replaced by etiher ``∈`` or ``=``

> The index variable is looped over for all integers from ``1:n``. 

Reviewer 2: ```suggestion
The index variable is looped over for all integers from ``1:n``
```

indices

Note that this does not actually create a vector of those. Instead, it creates an **iterator** that is looped over the **range** of integers from ``1`` to ``n``

> While in this example we specified the range of integers, there is a more memory efficient way:

Reviewer 2: ```suggestion
While in this example we specified the range of integers, there is a more memory efficient way
```

        end
        ϵ_mean = ϵ_sum / m

> 

Reviewer: I thought this was useful.  Not sure why you removed it?

        end
        ϵ_mean = ϵ_sum / m

> Of course, in Julia there are built in functions to perform this calculation which we can compare against:

Reviewer 2: ```suggestion
where ``ϵ[1:m]`` returns the elements of the vector at indices ``1`` to ``m``

Of course, in Julia there are built in functions to perform this calculation which we can compare against
```

    data = generatedata(10)
    plot(data)

Here:

* ``function`` is a Julia keyword that indicates the start of a function definition

* ``generatedata`` is an arbitrary name for the function

* ``return`` is a keyword indicating the return value

> By "remembering" that ``randn`` can return a vector, we could have a better code:

Reviewer 2: ```suggestion
By "remembering" that ``randn`` can return a vector, we could have a better code
```

    end
    data = generatedata(5)

While better, the looping over the ``i`` index to square the results is difficult to read

Instead of looping, we can instead **broadcast** the ``^2`` square function over a vector using a ``.``

To be clear, unlike Python, R, and Matlab (to a lesser extent), the reason to drop the ``for`` is **not** for performance reasons, but rather because of code clarity

Loops of this sort are at least as efficient than vectorized approach in compiled languages like Julia, so use a for loop if you think it makes the code more clear

> Furthermore, we can just drop the ``return`` because functions return the last calculation by default, which leads to:

Reviewer 2: ```suggestion
Furthermore, we can just drop the ``return`` because functions return the last calculation by default, which leads to
```

    generatedata(n) = randn(n).^2
    data = generatedata(5)    

> Finally, we can broadcast any function:

Reviewer 2: ```suggestion
Finally, we can broadcast any function
```

This function will be passed a choice of probability distribution and respond by plotting a histogram of observations

In doing so we'll make use of the Distributions package which was instantiated above with the project.

> Here's the code:

Reviewer 2: ```suggestion
Here's the code
```

The name ``lp`` is bound to this value

When we make the function call ``plothistogram(lp, 500)`` the code in the body

> of the function ``plothistogram`` is run with:

Reviewer 2: ```suggestion
of the function ``plothistogram`` is run with
```

So how can it be that ``rand()`` is able to take this kind of value as an
argument and return the output that we want?

> The answer in a nutshell is **multiple dispatch** which Julia uses to implement **generic programming**.

Reviewer 2: ```suggestion
The answer in a nutshell is **multiple dispatch** which Julia uses to implement **generic programming**
```

Example: Variations on Fixed-Points
================================================

For our second example, we will start with a simple example of determining fixed-points of a function

The goal is to start with a code similar to Matlab, and move towards a more **Julian** style with high mathematical clarity

Fixed-Point Maps
----------------------------

> Consider the simple equation, where the scalars :math:`p,\beta` are given, and  :math:`v` is the scalar we wish to solve for:

Reviewer 2: ```suggestion
Consider the simple equation, where the scalars :math:`p,\beta` are given, and  :math:`v` is the scalar we wish to solve for
```

    end
    println("Fixed point = $v_old, and |f(x) - x| = $normdiff in $iter iterations")

The ``while`` loop, like the ``for`` loop, should only be used directly in Jupyter or inside of a function

Here, we have used the ``norm`` function (from the ``LinearAlgebra`` base library) to compare the values

The other new function is the ``println`` with the string interpolation which splices a value of an expression or variable prefixed by ``$`` into a string

> An alternative approach is to use a ``for`` loop, and checking for convergence in each iteration:

Reviewer 2: ```suggestion
An alternative approach is to use a ``for`` loop, and checking for convergence in each iteration
```

    end
    println("Fixed point = $v_old, and |f(x) - x| = $normdiff in $iter iterations")

The new feature is ``break`` which leaves a ``for`` or a ``while`` loop

Using a Function
---------------------

> The first problem with this setup is that it depends on being sequently run. Therefore, it can be easily remedied with a function:

Reviewer 2: ```suggestion
The first problem with this setup is that it depends on being run in sequence

It can be easily remedied with a function
```

The chief issue is that the algorithm (finding a fixed point) is reusable and generic, while the function we calculate ``p + β * v`` is specific to our problem

> A key feature of languages like Julia is the ability to efficiently handle functions passed to other functions:

Reviewer 2: ```suggestion
A key feature of languages like Julia is the ability to efficiently handle functions passed to other functions
```

    sol = fixedpointmap(f, iv = 0.8, tolerance = 1.0E-8) # don't need to pass 
    println("Fixed point = $(sol.value), and |f(x) - x| = $(sol.normdiff) in $(sol.iter) iterations")

In this example, all function parameters after the ``;`` in the list must be called by name

> Furthermore, a default value may be enabled. So the named parameter ``iv`` is required while ``tolerance`` and ``maxiter`` have default values

Reviewer 2: ```suggestion
Furthermore, a default value may be enabled

So the named parameter ``iv`` is required while ``tolerance`` and ``maxiter`` have default values
```

> The return type of the function also has named fields, ``value, normdiff,`` and ``iter``. All accessed intuitively using ``.``

Reviewer 2: ```suggestion
The return type of the function also has named fields, ``value, normdiff,`` and ``iter``

All can be accessed intuitively using ``.``
```

Using a Package
----------------------------

> But best of all is to avoid writing code altogether:

Reviewer 2: ```suggestion
But best of all is to avoid writing code altogether
```

The ``fixedpoint`` function from the ``NLsolve.jl`` library implements the simple fixed-point iteration scheme above

Since the ``NLsolve`` library only accepts vector based inputs, we needed to make the ``f(v)`` function broadcast on the ``+`` sign, and pass in the initial condition as a vector of length 1 with ``[0.8]``

A key benefit of using a package is that the code is clearer, and the implementation is tested. Also, by using an orthogonal library we also enable performance improvements

> For instance, we can use the ``Anderson Acceleration`` (it is a method to calculate fixed points) with a memory of 5 iterations by changing a setting:

Reviewer 2: ```suggestion
The default is an algorithm called ``Anderson acceleration``, with a memory parameter of 5
```
